src/persist/src/indexed/mod.rs

```diff
@@ -689,11 +714,93 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             self.blob.delete_unsealed_batch(&batch)?;
         }
 
-        for batch in deleted_trace_batches {
-            self.blob.delete_trace_batch(&batch)?;
+        Ok(())
+    }
+
+    fn compact_trace_maintenance_reqs(&mut self) -> Result<Vec<MaintenanceReq>, Error> {
+        let mut reqs = vec![];
+        for (id, arrangement) in self.state.arrangements.iter() {
+            // Skip over the arrangement if it already has a compaction request in flight.
+            //
+            // TODO: note that this works because right now the connection between Indexed
+            // and Maintainer is infallible, and the two cannot fail independently of each
+            // other. If that changes, we will have to revisit the policy around waiting
+            // arbitrarily long for maintenance responses and most likely institute
+            // a timeout.
+            if self.in_flight_trace_compactions.contains_key(&id) {
+                continue;
+            }
+            if let Some(req) = arrangement.trace_next_compact_req()? {
+                reqs.push(MaintenanceReq::CompactTrace((*id, req.clone())));
+                self.in_flight_trace_compactions.insert(*id, req);
+            }
+        }
+        Ok(reqs)
+    }
+
+    fn do_handle_compact_trace_res(&mut self, id: Id, res: CompactTraceRes) {
+        let written_bytes = res.merged.size_bytes;
+        self.metrics.compaction_write_bytes.inc_by(written_bytes);
+
+        if !self.state.arrangements.contains_key(&id) {
+            log::trace!(
+                "received trace compaction response for deleted stream id: {:?}. Ignoring.",
+                id
+            );
+            return;
         }
 
-        Ok(())
+        self.apply_batched_cmd(|state, pending| {
+            let arrangement = state
+                .arrangements
+                .get_mut(&id)
+                .expect("arrangement known to exist");
+            let (_, deleted_trace_batches) = arrangement.trace_handle_compact_response(res);
+            if !deleted_trace_batches.is_empty() {
+                pending.add_deleted_trace_batches(id, deleted_trace_batches);
+            }
+        });
+    }
+
+    /// Handle the results of a previously sent maintenance request.
+    fn handle_maintenance_res(&mut self, res: MaintenanceRes) {
+        match res {
+            MaintenanceRes::CompactTrace((id, res)) => {
+                if !self.in_flight_trace_compactions.contains_key(&id) {
+                    log::trace!(
+                        "received trace compaction response for a stream that did not have a in flight request: {:?}. Ignoring",
+                        id
+                    );
+                    return;
+                }
+
+                match res {
```

Comment:
this match could use some early-return unnesting as well

Reply:
done.